The sample and empty flag (#6978)
dbt-labs/docs.getdbt.com: website/docs/docs/build/sample-flag.md

description: "Use the sample flag to lower development time and reduce warehouse spend."
sidebar_label: "The sample flag"
pagination_next: null
pagination_prev: "docs/build/empty-flag"
---

# About the `--sample` flag

Large data sets can drastically increase build times and reduce how quickly dbt developers can build and test new code. The dbt `--sample` flag can help to reduce build times when developing and warehouse spend by running dbt in sample mode. Sample mode enables you to address cases where you don't need to build the entire model during the development or CI cycle but include enough data to validate the outputs. 

Sample mode builds upon the [`--empty` flag](/docs/build/empty-flag) by taking the validation of semantic results a step further by including a sampling of data from the model(s) in your development schema. It won't solve every scenario; for example, there are cases where not all joins will be populated. However, it presents a viable solution for faster building, testing, and validating many strategies. 

The `--sample` flag will become more robust over time, but it only supports time-based sampling for now.

## Using the `--sample` flag

> The `--sample` flag is available for the [`run`](/reference/commands/run) and [`build`](/reference/commands/build) commands, and when used, sample mode generates filtered refs and sources. Since it's using time-based sampling, if you have refs like {{ ref('some_model') }} being sampled, you need to set [`event_time`](/reference/resource-configs/event-time) for `some_model` to the field that will be used as the timestamp. 

Author: ```suggestion
The `--sample` flag is available for the [`run`](/reference/commands/run) and [`build`](/reference/commands/build) commands, and when used, sample mode generates filtered refs and sources. Since it's using time-based sampling, if you have refs like `{{ ref('some_model') }}` being sampled, you need to set [`event_time`](/reference/resource-configs/event-time) for `some_model` to the field that will be used as the timestamp. 
```

### Example

Let's say you want to run your `stg_customers` model and build the table in your development schema with a sample size of three days. Your command in the IDE would look something like this:

```
dbt run --select path/to/stg_customers --sample --time="3 days"
```

If you have an even larger model, for example, `stg_orders` you can set sample mode to hours:

```
dbt run --select path/to/stg_customers --sample --time="6 hours"
```

> dbt will then execute the model SQL against the target data warehouse and build the tables with data from the sample size. 

Reviewer: Hey sorry for giving you no guidance on this, and thank you for putting this together despite that! ❤️ 

One thing we need to fix is that there is no `--time` flag now, instead all one needs to do is `--sample="6 hours"`.

There are additionally some details we should add. In the example `--sample="6 hours"`, the `"6 hours"` part is what we call the "sample specification". Presently there are two time based sample specifications supported, relative time specs and static time specs. 

The one we've already seen is a "relative time spec", that comes in the form `<int> <granularity>`, and makes your sample filter from `now` to `now - (<int> * <granularity>)`. Supported granularities currently are `hours`, `days`, `months`, and `years`.

Static time specs look like `--sample="{'start': '2025-01-03', 'end': '2025-01-07 12:34:56'}"`. With them you can define a specific `start` and `end` of your sample time filter. Note that the start is "inclusive" and the end is "exlcusive".

---
title: "About the sample flag"
description: "Use the sample flag to lower development time and reduce warehouse spend."
sidebar_label: "The sample flag"
pagination_next: null
pagination_prev: "docs/build/empty-flag"
---

# About the `--sample` flag

:::note

> The `--sample` flag is not currently available for Python models. If the flag is used with a Python model, it will be ignored.

Reviewer 2: @QMalcolm confirming this is also the case for snapshots ... and seeds?

Reviewer: Seeds can be sampled. I.e. `ref(my_seed)` will sample the seed. A seed itself cannot be created as a sample of its inputs because it doesn't actually have any inputs to sample, it is the root node.

For snapshots, during the creation of a snapshot, its inputs will be sampled if running in sample mode. Also, a snapshot _can be_ sampled as an input to another node.

Author: Do we want to specify that out in the body of the page or does the callout need to be used to address it?

Reviewer: @matthewshaver I don't think we need to callout anything with regard to snapshots, because they'll just work. @graciegoheen does that sound correct?

As for seeds, I think a callout of something like the following would be nice:
> Seeds will be created as normal, but they will sampled when referenced but down stream nodes

:::

> Large data sets can drastically increase build times and reduce how quickly dbt developers can build and test new code. The dbt `--sample` flag can help to reduce build times when developing and warehouse spend by running dbt in sample mode. Sample mode enables you to address cases where you don't need to build the entire model during the development or CI cycle but include enough data to validate the outputs. 

Reviewer 3: ```suggestion
Large data sets can drastically increase build times and reduce how quickly dbt developers can build and test new code. The dbt `--sample` flag can help to reduce build times and warehouse spend by running dbt in sample mode. Sample mode enables you to address cases where you don't need to build the entire model during the development or CI cycle but include enough data to validate the outputs. 
```

> Sample mode builds upon the [`--empty` flag](/docs/build/empty-flag) by taking the validation of semantic results a step further by including a sampling of data from the model(s) in your development schema. It won't solve every scenario; for example, there are cases where not all joins will be populated. However, it presents a viable solution for faster building, testing, and validating many strategies. 

Author: I'm not sure I'd use "refines" here as that would indicate the empty flag is inferior or incomplete when they just serve two different purposes. How about:

```suggestion
Sample mode takes the [`--empty` flag's](/docs/build/empty-flag) validation of semantic results a step further by including a sampling of data from the model(s) in your development schema. It won't solve every scenario; for example, there are cases where not all joins will be populated. However, it presents a viable solution for faster building, testing, and validating many strategies. 
```

Reviewer 3: Looks great matt!

> Sample mode builds upon the [`--empty` flag](/docs/build/empty-flag) by taking the validation of semantic results a step further by including a sampling of data from the model(s) in your development schema. It won't solve every scenario; for example, there are cases where not all joins will be populated. However, it presents a viable solution for faster building, testing, and validating many strategies. 

Reviewer 3: Replacing "builds" as it's used a lot in our docs 😬 

Should we also mention when joins will not be populated? which cases?

```suggestion
Sample mode refines the [`--empty` flag](/docs/build/empty-flag), taking the validation of semantic results a step further by including a sampling of data from the model(s) in your development schema. It won't solve every scenario; for example, there are cases where not all joins will be populated. However, it presents a viable solution for faster building, testing, and validating many strategies. 
```

Author: I'm not sure I'd use "refines" here as that would indicate the empty flag is inferior or incomplete when they just serve two different purposes. How about:

The `--sample` flag will become more robust over time, but it only supports time-based sampling for now.

## Using the `--sample` flag

> The `--sample` flag is available for the [`run`](/reference/commands/run) and [`build`](/reference/commands/build) commands, and when used, sample mode generates filtered refs and sources. Since it's using time-based sampling, if you have refs like `{{ ref('some_model') }}` being sampled, you need to set [`event_time`](/reference/resource-configs/event-time) for `some_model` to the field that will be used as the timestamp. 

Reviewer 3: ```suggestion
The `--sample` flag is available for the [`run`](/reference/commands/run) and [`build`](/reference/commands/build) commands. When used, sample mode generates filtered refs and sources. Since it's using time-based sampling, if you have refs like `{{ ref('some_model') }}` being sampled, you need to set [`event_time`](/reference/resource-configs/event-time) for `some_model` to the field that will be used as the timestamp. 
```

There are two time-based sample specifications supported for sample mode:
- **Relative time specs:** Filters sampled data from the time the command is run back to a specified integer and granularity. Supported granularities are:
    - Hours
    - Days
    - Months
    - Years

> - **Static time specs:** Filters your data between a defined start and end period

Author: ```suggestion
- **Static time specs:** Filters your data between a defined start and end period using date and/or timestamp.
```

> ### Example

Reviewer 3: ```suggestion
### Examples
```

Let's say you want to run your `stg_customers` model and build the table in your development schema with a relative time spec sample size of three days. Your command in the IDE would look something like this:

```
dbt run --select path/to/stg_customers --sample="3 days"
```

If you have an even larger model, for example, `stg_orders` you can set sample mode to hours:

```
dbt run --select path/to/stg_customers --sample="6 hours"
```

Next, let's say you want to validate data for your entire business from a sample size further in the past - your busiest week in July, from the first until closing time on the eighth. You can run the following: 

```
dbt run --sample="{'start': '2024-07-01', 'end': '2024-07-08 18:00:00'}"
```

dbt will then execute the model SQL against the target data warehouse and build the tables with data from the sample sizes.

>  

Reviewer 2: @QMalcolm thoughts on adding a note about using .render() to make it so a given ref/source will _not_ be sampled?

Reviewer: Yea that would be a good thing to include 🤔

Author: Would you mind a little guidance on how we want to address this?

Reviewer: @matthewshaver Something along the lines of
> If you would like a ref to never be sampled when running with `--sample` then append `.render()` to the `ref`, e.g. `{{ ref("my_model").render() }}`

Author: Thanks @QMalcolm - Added to the doc!
